### ISSUE TITLE
Torch BF16 group gemm hangs in backward pass - core issue isolated, needs proper resolution.

### ISSUE THREAD

**Issue by lessw2020** [pytorch/pytorch#152668]
### 🐛 Describe the bug

When using the native torch.__group_gemm enabled via https://github.com/pytorch/pytorch/pull/150374,
users reported hanging after a certain number of iterations in the backward pass. (https://github.com/pytorch/torchtitan/issues/1118)
With a deeper dive from Razvan, a min repro case has been established that points to the core issue being that if an m_offsets is empty (i.e an expert has no assigned tokens) then torch gg will be fine in forward, but hang in backwards.  
We find that padding any empty offset with at least 8, avoids the hang. 
From there, additional testing was done b/c users also reported that Adam vs AdamW made a difference in the hang with Adam running for longer.
Using titan llama4 and running with Adam, I verified that the hang occurs as soon as a given expert encounters a zero token's assigned. 
By contrast, when running with AdamW, this hang is ultimately encountered much sooner b/c there, an expert only has to get to the min_aligment_m used in offset creation - tested and verified with both 8 and 16 as min_alignment in the create permute indices.  
As soon as an expert hits that number of tokens, it will hang there rather than getting to zero like with Adam and thus matches user reported differences based on optimizer. 

example with Adam and zero - screenshot of hang:

<img width="1016" alt="Image" src="https://github.com/user-attachments/assets/5fc1637c-d97e-40bd-9beb-f77a0ef32b1c" />

by contrast with AdamW and min_alignment of 8 - screenshot of hang and token to expert assignment...note that 2 have 8 assigned:

<img width="1016" alt="Image" src="https://github.com/user-attachments/assets/579483ff-f884-4893-b748-441f1946f076" />

Easy repro scenario:

~~~
import torch

num_experts = 4
M, K, N = 48, 8, 16

# to repro hang, make a given expert have 0 tokens ala (0, 8, 16, 32, 40) or (8,8,16,32,40)
m_offsets_hang = (8, 8, 32, 40)
m_offsets = (8, 16, 32, 40)
x = torch.randn(M, K, dtype=torch.bfloat16, device="cuda", requires_grad=True)
print(f"{x.shape=}")
w = torch.randn(
    num_experts, K, N, dtype=torch.bfloat16, device="cuda", requires_grad=True
)
print(f"{w.shape=}")
offs = torch.tensor(m_offsets, dtype=torch.int32, device="cuda")


print(f"Running simple forward...")
o = torch._grouped_mm(x, w, offs)
print(f"forward completed!")
print(f"Running backward...")
o.backward(torch.randn_like(o))
print(f"backward completed!")
torch.cuda.synchronize()
print(f"Completed! {o.shape=}")
~~~

Probably two resolutions here:
a - for implementation, will work on padding out any empty m_offsets to avoid passing in zero. 
b - however, the kernel itself likely needs to also check this, or at least we need to document that it requires no empty offsets? 
c - unclear what difference the optimizer is making but clearly there is a subtle difference as shown above.  However, maybe we don't care if this all goes away with min padding and enforcement of no zero offsets. 

### Versions

latest nightly will work fine to run min repro. 
Use current torchtitan and llama4 in experimental to repro there. 

for completeness:
PyTorch version: 2.8.0.dev20250430+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-5)
Clang version: Could not collect
CMake version: version 3.26.5
Libc version: glibc-2.34

Python version: 3.12.9 | packaged by Anaconda, Inc. | (main, Feb  6 2025, 18:56:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.4.3-0_fbk20_zion_2830_g3e5ab162667d-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.6.85
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100
GPU 1: NVIDIA H100
GPU 2: NVIDIA H100
GPU 3: NVIDIA H100
GPU 4: NVIDIA H100
GPU 5: NVIDIA H100
GPU 6: NVIDIA H100
GPU 7: NVIDIA H100

Nvidia driver version: 535.154.05
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.9.8.0
/usr/lib64/libcudnn_adv.so.9.8.0
/usr/lib64/libcudnn_cnn.so.9.8.0
/usr/lib64/libcudnn_engines_precompiled.so.9.8.0
/usr/lib64/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/lib64/libcudnn_graph.so.9.8.0
/usr/lib64/libcudnn_heuristic.so.9.8.0
/usr/lib64/libcudnn_ops.so.9.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             384
On-line CPU(s) list:                0-383
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 9654 96-Core Processor
CPU family:                         25
Model:                              17
Thread(s) per core:                 2
Core(s) per socket:                 96
Socket(s):                          2
Stepping:                           1
Frequency boost:                    enabled
CPU(s) scaling MHz:                 84%
CPU max MHz:                        3707.8120
CPU min MHz:                        1500.0000
BogoMIPS:                           4792.85
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d
Virtualization:                     AMD-V
L1d cache:                          6 MiB (192 instances)
L1i cache:                          6 MiB (192 instances)
L2 cache:                           192 MiB (192 instances)
L3 cache:                           768 MiB (24 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-95,192-287
NUMA node1 CPU(s):                  96-191,288-383
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Vulnerable: eIBRS with unprivileged eBPF
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==2.2.3
[pip3] nvidia-cublas-cu12==12.8.3.14
[pip3] nvidia-cuda-cupti-cu12==12.8.57
[pip3] nvidia-cuda-nvrtc-cu12==12.8.61
[pip3] nvidia-cuda-runtime-cu12==12.8.57
[pip3] nvidia-cudnn-cu12==9.8.0.87
[pip3] nvidia-cufft-cu12==11.3.3.41
[pip3] nvidia-curand-cu12==10.3.9.55
[pip3] nvidia-cusolver-cu12==11.7.2.55
[pip3] nvidia-cusparse-cu12==12.5.7.53
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.8.61
[pip3] nvidia-nvtx-cu12==12.8.55
[pip3] pytorch-triton==3.3.0+git96316ce5
[pip3] torch==2.8.0.dev20250430+cu128
[pip3] torchao==0.11.0+git2fcab01d
[pip3] torchdata==0.11.0
[pip3] torchtitan==0.0.2
[pip3] triton==3.3.0
[conda] numpy                     2.2.3                    pypi_0    pypi
[conda] nvidia-cublas-cu12        12.8.3.14                pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.8.57                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.8.61                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.8.57                  pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.8.0.87                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.3.3.41                pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.9.55                pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.7.2.55                pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.5.7.53                pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.3                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.26.2                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.8.61                  pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.8.55                  pypi_0    pypi
[conda] pytorch-triton            3.3.0+git96316ce5          pypi_0    pypi
[conda] torch                     2.8.0.dev20250430+cu128          pypi_0    pypi
[conda] torchao                   0.11.0+git2fcab01d           dev_0    <develop>
[conda] torchdata                 0.11.0                   pypi_0    pypi
[conda] torchtitan                0.0.2                    pypi_0    pypi
[conda] triton                    3.3.0                    pypi_0    pypi

**Comment by lessw2020** [comment]
cc @jwfromm @kwen2501 @ngimel @danielvegamyhre @drisspg @tianyu-l @ebsmothers

**Comment by ngimel** [comment]
It's a known issue with cutlass, they promise a fix for this. My attempt at adding a kernel assert for K=0 to at least avoid hang hit a problem with torch.compile autotuning of grouped gemm - something in that path runs fp8 gouped gemm with K=0 for all groups, and for some reason that doesn't hang so the issue was not identified earlier. I'm working with @bertmaher to make sure that we don't run K=0 under normal circumstances, and I'll put a device assert for K=0 into grouped gemm once that's resolved

**Comment by lessw2020** [comment]
have a PR open now to resolve for llama4 / deepseek usage in Titan:
https://github.com/pytorch/torchtitan/pull/1166

This is effectively an auto-padding for experts with zero tokens.   Now able to run to 2K iters with no issues.

**Comment by bertmaher** [comment]
Merging https://github.com/pytorch/pytorch/pull/152968 which should solve the issue with autotuning generating K=0 grouped gemms.  Sorry for the oversight!

**Comment by jcao-ai** [comment]
Thanks, it works like a charm.
